### PR TITLE
Add API for deserializing data envelopes from JSON

### LINKED ISSUE #391: Make broker::data_envelope::deserialize() for JSON format accessible
I have found that I can parse Zeek events from arbitrary v1 binary format byte buffers with the following snippet:
```
    auto r = broker::data_envelope::deserialize(broker::endpoint_id::nil(), broker::endpoint_id::nil(), 0, "", payload,
                                                payload_size);
    if ( ! r )
        return std::nullopt;

    broker::zeek::Event ev(std::move(*r));
```

I have also found that I should be able to serialize Zeek events into v1 JSON byte buffers as follows:
```
    auto ev = broker::zeek::Event(event.HandlerName(), std::move(xs), broker::to_timestamp(event.timestamp));
    broker::format::json::v1::encode(ev.move_data(), std::back_inserter(buf));
```

Going from JSON to back to a broker::zeek::Event however isn't accessible unless I've missed something. I've found  the following place though:

https://github.com/zeek/broker/blob/861c20612e8b28811458ff9a96f091cc6de0c005/src/internal/json_client.cc#L126-L142
 
Could this be lifted into an API so that de-serialization can be re-used outside of json_client? Getting back the broker::data_message_ptr after the `de-re-serialization` step would be great.

Background is re-using the same serialization format over a different transport, which works for the binary format.

### PR DESCRIPTION
**Pull request by Neverlord**
Closes #391.

### CODE REVIEW

**Comment by Neverlord** [comment]
Reverting back to draft, it doesn't deserialize the topic from the data yet.

**Comment by Neverlord** [comment]
Ready to go now.

File: libbroker/broker/envelope.cc [hunk 1]
```diff
@@ -148,6 +152,39 @@ expected<envelope_ptr> envelope::deserialize(const std::byte* data,
...
 
+expected<envelope_ptr> envelope::deserialize_json(const char* data,
+                                                  size_t size) {
+  // Parse the JSON text into a JSON object.
+  auto val = caf::json_value::parse_shallow(std::string_view{data, size});
+  if (!val)
+    return error{ec::invalid_json};
+  auto obj = val->to_object();
+  // Type-checking.
+  if (obj.value("type").to_string() != "data-message")
```
Comment (awelzel) [inline libbroker/broker/envelope.cc:163]: Hmm, I realize I'm looking at parsing one layer further in: The idea was, JSON contains only the serialized event vector without the `data-message` and `topic`.

In my use-case, "type" and "topic"  are in message headers, not within the payload and the payload isn't meant to be structured like the websocket protocol.

Let me check if the JSON serialization in #391 even does what I thought it did.

Comment (Neverlord) [reply libbroker/broker/envelope.cc:163]: So you'd basically want to have a function that parses you just the `variant` from the JSON format? I'll add that later.

Comment (Neverlord) [reply libbroker/broker/envelope.cc:163]: > Now, the main motivation is to "just" re-use existing encoding/serialization functionality from broker for events. If you think that API doesn't make much sense to have, feel free to push back.

No push back at all, I just didn't think of that use case yet. 🙂

I'll add functions that do the JSON (de)-serialization for `variant` as well.

**Comment by Neverlord** [comment]
All done.

**Review by timwoj — APPROVED** [review]

**Comment by timwoj** [comment]
I reviewed this a while ago, and just realized that I never approved it.

**Pull request merged**